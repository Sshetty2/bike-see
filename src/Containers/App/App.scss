*,
*:after,
*:before {
  box-sizing: border-box;
  font-family: 'Quicksand', sans-serif;
}

body {
  margin: 0;
  padding: 0;
}

i {
  font-size: 24px;
  font-family: Font Awesome 5 Free;
  margin: 10px;
}