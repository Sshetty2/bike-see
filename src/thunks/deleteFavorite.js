import { removeFavorite, setLoading, setError } from '../actions';
import { fetchData } from '../utils/api';

export const deleteFavorite = (station_id, user_id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      await fetchData(`https://bike-see.herokuapp.com/api/v1/users/${user_id}/favorites/${station_id}`, 'DELETE');
      dispatch(removeFavorite(station_id));
    } catch (error) {
      dispatch(setError('Error deleting favorite.'));
    }
    dispatch(setLoading(false));
  }
}