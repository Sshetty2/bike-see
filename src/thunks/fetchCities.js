import { setLoading, setError, setCities } from '../actions';
import { fetchData } from '../utils/api';

export const fetchCities = () => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const results = await fetchData('https://api.citybik.es/v2/networks', 'GET');
      dispatch(setCities(results.networks));
    } catch (error) {
      dispatch(setError('Error fetching cities.'));
    }
    dispatch(setLoading(false));
  }
}