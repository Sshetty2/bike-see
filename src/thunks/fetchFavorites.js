import { setFavorites, setLoading, setError } from '../actions';
import { fetchData } from '../utils/api';

export const fetchFavorites = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const favorites = await fetchData(`http://localhost:3001/api/v1/users/${id}/favorites`, 'GET');
      const favoriteStations = favorites.map(favorite => favorite.station_id);
      dispatch(setFavorites(favoriteStations));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}
  