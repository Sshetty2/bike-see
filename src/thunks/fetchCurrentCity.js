import { setLoading, setError, setCurrentCity } from '../actions';
import { fetchStations } from '../thunks/fetchStations';
import { fetchData } from '../utils/api';

export const fetchCurrentCity = (user_id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const fetchedCity = await fetchData(`http://localhost:3001/api/v1/users/${user_id}/city`, 'GET');

      if (fetchedCity !== '') {
        dispatch(setCurrentCity(fetchedCity.city));
        dispatch(fetchStations(user_id, fetchedCity.city));
      }

    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}