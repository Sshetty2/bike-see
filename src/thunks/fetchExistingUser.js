import { setCurrentUser, setLoading, setError } from '../actions';
import { fetchFavorites } from '../thunks/fetchFavorites';
import { fetchData } from '../utils/api';

export const fetchExistingUser = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const fetchedUser = await fetchData(`/api/v1/users/${id}`, 'GET');
      dispatch(setCurrentUser(fetchedUser));
      dispatch(fetchFavorites(fetchedUser.id));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}