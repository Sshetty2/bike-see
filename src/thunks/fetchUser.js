import { setCurrentUser, setLoading, setError } from '../actions';
import { fetchCurrentCity } from '../thunks/fetchCurrentCity';
import { fetchFavorites } from '../thunks/fetchFavorites';
import { fetchData } from '../utils/api';

export const fetchUser = (user) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const fetchedUser = await fetchData('http://localhost:3001/api/v1/users', 'POST', user);
      dispatch(setCurrentUser(fetchedUser));
      localStorage.setItem('bike-user', JSON.stringify(fetchedUser));
      dispatch(fetchFavorites(fetchedUser.id));
      dispatch(fetchCurrentCity(fetchedUser.id));
    } catch (error) {
      dispatch(setError('Error logging in user. Please sign up or try again.'));
    }
    dispatch(setLoading(false));
  }
}