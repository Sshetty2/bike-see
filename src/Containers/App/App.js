import React, { Component } from 'react';
import Header from '../Header/Header';
import BikeMap from '../BikeMap/BikeMap';
import PopUp from '../../Components/PopUp/PopUp';
import { Route, Switch, withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { fetchStations } from '../../thunks/fetchStations';
import { fetchCities } from '../../thunks/fetchCities';
import { fetchFavorites } from '../../thunks/fetchFavorites';
import PropTypes from 'prop-types';
import { fetchUser } from '../../thunks/fetchUser';

export class App extends Component {

  getCurrentUser = () => {
    const { fetchUser, fetchFavorites } = this.props;
    if (localStorage.hasOwnProperty('bike-user')) {
      const savedUser = localStorage.getItem('bike-user');
      const parsedUser = JSON.parse(savedUser);
      fetchUser(parsedUser);
      fetchFavorites(parsedUser.id);
      fetchStations();
    }
  }

  componentDidMount() {
    this.props.fetchCities();
    this.getCurrentUser();
  }

  render() {
    return (
      <div id='app-div'>
        <Header />
        <Switch>
          <Route path='/' component={BikeMap} />
          <Route path='/stations' component={BikeMap} />
          <Route path='/my-stops' component={BikeMap} />
        </Switch>
        <Route path="/login" render={() => <PopUp history={this.props.history} title='User Login'/>} />
        <Route path="/sign-up" render={() => <PopUp history={this.props.history} title='User Sign Up' />} />
      </div>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  fetchCities: (cities) => dispatch(fetchCities(cities)),
  fetchStations: (stations) => dispatch(fetchStations(stations)),
  fetchFavorites: (user) => dispatch(fetchFavorites(user)),
  fetchUser: (user) => dispatch(fetchUser(user)),
});

App.propTypes = {
  fetchCities: PropTypes.func,
  fetchStations: PropTypes.func,
  fetchFavorites: PropTypes.func,
  fetchUser: PropTypes.func,
}

export default withRouter(connect(null, mapDispatchToProps)(App));
