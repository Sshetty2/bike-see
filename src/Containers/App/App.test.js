import React from 'react';
import { App, mapDispatchToProps } from './App';
import { shallow } from 'enzyme';
import { fetchCities } from '../../thunks/fetchCities';
import { fetchUser } from '../../thunks/fetchUser';
jest.mock('../../thunks/fetchCities');
jest.mock('../../thunks/fetchUser');

describe('App', () => {
  let wrapper;
  let fetchCitiesMock;
  let fetchUserMock;

  beforeEach(() => {
    fetchCitiesMock = jest.fn();
    fetchUserMock = jest.fn();
    wrapper = shallow(
      <App
        fetchCities={fetchCitiesMock}
        fetchUser={fetchUserMock}
      />
    )
  });

  it('should match the correct snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('componentDidMount', () => {
    it('should call fetchCities', () => {
      wrapper.instance().componentDidMount();
      expect(wrapper.instance().props.fetchCities).toHaveBeenCalled();
    });

    it('should call getCurrentUser', () => {
      wrapper.instance().getCurrentUser = jest.fn();
      wrapper.instance().componentDidMount();
      expect(wrapper.instance().getCurrentUser).toHaveBeenCalled();
    });
  });

  describe('getCurrentUser', () => {
    it('should fetch user if it exists in local storage', () => {
      const mockUser = { id: '1', name: 'Karin', email: 'k@k', password: 'password'};
      localStorage.setItem('bike-user', JSON.stringify(mockUser));
      wrapper.instance().getCurrentUser();
      expect(wrapper.instance().props.fetchUser)
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch when fetchCities is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchCities();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchCities();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch when fetchUser is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchUser();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchUser();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});