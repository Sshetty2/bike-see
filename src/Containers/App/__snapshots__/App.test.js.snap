// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the correct snapshot 1`] = `
<div
  id="app-div"
>
  <withRouter(Connect(Header)) />
  <Switch>
    <Route
      component={[Function]}
      path="/not-found"
    />
    <Route
      component={[Function]}
      path="/"
    />
  </Switch>
  <Switch>
    <Route
      component={[Function]}
      path="/stations"
    />
    <Route
      component={[Function]}
      path="/my-stops"
    />
    <Route
      path="/login"
      render={[Function]}
    />
    <Route
      path="/sign-up"
      render={[Function]}
    />
    <Route
      path="/:anythingelse"
      render={[Function]}
    />
  </Switch>
</div>
`;
